base/sort.jl

```diff
@@ -761,6 +761,12 @@ function _extrema(v::AbstractVector, lo::Integer, hi::Integer, o::Ordering)
     end
     mn, mx
 end
+function _issorted(v::AbstractVector, lo::Integer, hi::Integer, o::Ordering)
+    @inbounds for i in (lo+1):hi
```

Comment:
Is `@inbounds` safe here? I didn't see a check for `lo` and `hi` being in `eachindex(v)`, but perhaps I've missed it. Maybe that needs to be added if it's not there already

Reply:
Sort of. The safety issues here are part of the larger issue of clearly demarcating the sorting API. [Officially](https://docs.julialang.org/en/v1/manual/faq/#man-api), "Functions, types, and constants are not part of the public API if they are not included in the documentation, _even if they have docstrings_." `_issorted`, much like `_extrema` above is clearly internal. `sort!(v::AbstractVector)` is clearly a part of the public API, and, as far as I can tell, whenever that public facing `sort!` calls `sort!(v, lo, hi, alg, order)`, `lo` and `hi` are drawn from within the axes of the input vector. The `sort!(v, lo, hi, alg, order)` method is not mentioned in the documentation and assumes `lo` and `hi` are inbounds everywhere already. This is bad because the function is public facing and it's ambiguous whether those methods are public or not.

If you call standard sorting methods it is somewhat safe, checked two-ish frames up the call stack. This is not good enough, but adding bounds checking here would not meaningfully improve the situation. For example, we make the same assumption that `lo` and `hi` are inbounds just above with `lenm1 < 40 && return sort!(v, lo, hi, SMALL_ALGORITHM, o)` and just below with `v_min, v_max = _extrema(v, lo, hi, o)`.

I plan to fix all this in a refactoring described [here](https://github.com/JuliaLang/julia/discussions/44876#discussioncomment-2768446) (hopefully in v1.9 or v1.10).